channeldb/waitingproof.go

```diff
@@ -76,35 +77,54 @@ func (s *WaitingProofStore) Add(proof *WaitingProof) error {
 		}
 
 		key := proof.Key()
-		if err := bucket.Put(key[:], b.Bytes()); err != nil {
-			return err
-		}
 
-		s.cache[proof.Key()] = struct{}{}
-		return nil
+		return bucket.Put(key[:], b.Bytes())
 	})
+	if err != nil {
+		return err
+	}
+
+	// Knowing that the write succeeded, we can now update the in-memory
+	// cache with the proof's key.
+	s.cache[proof.Key()] = struct{}{}
+
+	return nil
 }
 
 // Remove removes the proof from storage by its key.
 func (s *WaitingProofStore) Remove(key WaitingProofKey) error {
+	s.mu.RLock()
 	if _, ok := s.cache[key]; !ok {
+		s.mu.RUnlock()
 		return ErrWaitingProofNotFound
 	}
+	s.mu.RUnlock()
 
-	return s.db.Batch(func(tx *bolt.Tx) error {
+	s.mu.Lock()
```

Comment:
I can see that using a read-lock to check if it is worth acquiring the write-lock is an optimization, but is it worth the extra code complexity? Doesn't seem very performance critical tbh